core/query/delateralize.ml

```diff
@@ -0,0 +1,247 @@
+(*****************************************************************************
+ ** Delateralize.ml - Implements the delateralization algorithm             **
+ **                                                                         **
+ ** author: Wilmer Ricciotti                                                **
+ *****************************************************************************)
+open Utility
+open CommonTypes
+
+module Q = MixingQuery.Lang
+
+let (>>=) (o : 'a option) (f : 'a -> 'b option) =
+    match o with
+    | Some a -> f a
+    | _ -> None
+
+let (||=) o o' =
+    match o with
+    | None -> o'
+    | _ -> o
+
+(* let squash o = o >>= fun x -> x *)
+
+(* conjunctive bind for option lists, not used
+let rec (>>>=) (l : ('a option) list) (f : 'a -> 'b option) : ('b list) option =
+    match l with
+    | [] -> Some []
+    | o::ol ->
+           o >>= fun a ->
+           f a >>= fun b ->
+           (ol >>>= f) >>= fun bl ->
+           Some (b :: bl)
+*)
+
+let unopt_default ox x' = match ox with None -> x' | Some x'' -> x''
+
+let rec (>>==) (l : 'a list) (f : 'a -> 'a option) : 'a list option =
+    match l with
+    | [] -> None
+    | a::al ->
+        match f a, al >>== f with
+        | None, None -> None
+        | fa, fal ->
+            Some (unopt_default fa a::unopt_default fal al)
+
+(* not the most efficient way of doing these two functions: why is this not in the standard library? *)
```

Comment:
Yes. My efficiency complaint is based on the fact that when we find an element `x` of the collection for which `f x` is defined, we don't need to evaluate `f` on the remaining elements, nor do we need to scan the rest of the collection.

For lists, I can achieve that by ad-hoc recursion rather than using a fold. The same won't work on StringMap.t because its implementation is hidden: the best "standard" way I have to avoid evaluating `f` unnecessarily is by means of `lazy`/`force`, but `StringMap.fold` will still scan the whole map no matter what. Exceptions can be used to efficiently exit a loop in such a case, but since I need to extract a value of generic type `'a option`, I can't use exceptions without resorting to `Obj.(black)magic`, which I prefer to avoid.